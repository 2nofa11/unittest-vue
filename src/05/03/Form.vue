<script setup lang="ts">
defineProps<{ name: string }>();

const emit = defineEmits<{ (e: "submit", event: Event): void }>();
const handleSubmit = (event: Event) => {
  emit("submit", event);
};
</script>

<template>
  <form @submit.prevent="handleSubmit">
    <h2>アカウント情報</h2>
    <p>{{ name }}</p>
    <div>
      <button>編集する</button>
    </div>
  </form>
</template>
