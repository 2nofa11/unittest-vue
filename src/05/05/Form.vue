<script setup lang="ts">
import { ref } from "vue";
import Agreement from "./Agreement.vue";
import InputAccount from "./InputAccount.vue";

const email = ref("");
const password = ref("");
const checked = ref(false);
</script>

<template>
  <form @submit.prevent="handleSubmit" data-test="signup-form">
    <h2>新規アカウント登録</h2>
    <InputAccount v-model:email="email" v-model:password="password" />
    <Agreement v-model="checked" data-test="agreement-component" />
    <div>
      <button :disabled="!checked" data-test="submit-button">
        サインアップ
      </button>
    </div>
  </form>
</template>
