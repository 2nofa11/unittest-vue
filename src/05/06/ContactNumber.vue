<script setup lang="ts">
const phoneNumber = defineModel<string>("phoneNumber");
const name = defineModel<string>("name");
</script>

<template>
  <fieldset data-test="contact-fieldset">
    <legend data-test="contact-legend">連絡先</legend>
    <div>
      <label data-test="contact-number-label">
        電話番号
        <input
          type="tel"
          v-model="phoneNumber"
          data-test="contact-number-input"
        />
      </label>
    </div>
    <div>
      <label data-test="contact-name-label">
        お名前
        <input type="text" v-model="name" data-test="contact-name-input" />
      </label>
    </div>
  </fieldset>
</template>
