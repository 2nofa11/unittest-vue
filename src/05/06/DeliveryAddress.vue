<script setup lang="ts">
// import { prefectures } from "./fixtures"; // Removed import

const props = withDefaults(defineProps<{ title?: string }>(), {
  title: "お届け先",
});

const postalCode = defineModel<string>("postalCode");
const prefecture = defineModel<string>("prefecture");
const municipalities = defineModel<string>("municipalities");
const streetAddress = defineModel<string>("streetAddress");
</script>

<template>
  <fieldset data-test="delivery-address-fieldset">
    <legend data-test="delivery-address-legend">{{ props.title }}</legend>
    <div>
      <label data-test="postal-code-label">
        郵便番号
        <input
          type="text"
          placeholder="167-0051"
          v-model="postalCode"
          data-test="postal-code-input"
        />
      </label>
    </div>
    <div>
      <label data-test="prefecture-label">
        都道府県
        <input
          type="text"
          placeholder="東京都"
          v-model="prefecture"
          data-test="prefecture-input"
        />
      </label>
    </div>
    <div>
      <label data-test="municipalities-label">
        市区町村
        <input
          type="text"
          placeholder="杉並区荻窪1"
          v-model="municipalities"
          data-test="municipalities-input"
        />
      </label>
    </div>
    <div>
      <label data-test="street-address-label">
        番地番号
        <input
          type="text"
          placeholder="00-00"
          v-model="streetAddress"
          data-test="street-address-input"
        />
      </label>
    </div>
  </fieldset>
</template>
