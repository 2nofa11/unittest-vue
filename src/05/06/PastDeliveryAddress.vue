<script setup lang="ts">
const props = defineProps<{
  disabled?: boolean;
  options: AddressOption[];
}>();

const selectedAddress = defineModel<string>(); // Model for the selected address value
</script>

<template>
  <fieldset
    :disabled="props.disabled"
    :style="{ opacity: props.disabled ? 0.3 : 1 }"
    data-test="past-delivery-address-fieldset"
  >
    <legend data-test="past-delivery-address-legend">過去のお届け先</legend>
    <select
      v-model="selectedAddress"
      name="pastDeliveryAddress"
      data-test="past-delivery-address-select"
    >
      <option
        v-for="option in props.options"
        :key="option.id"
        :value="option.value"
        data-test="past-delivery-address-option"
      >
        {{ option.children }}
      </option>
    </select>
  </fieldset>
</template>
