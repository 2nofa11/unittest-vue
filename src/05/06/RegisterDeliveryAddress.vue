<script setup lang="ts">
const register = defineModel<boolean>();
</script>

<template>
  <fieldset data-test="register-delivery-address-fieldset">
    <legend data-test="register-delivery-address-legend">
      新しいお届け先を登録しますか？
    </legend>
    <label data-test="register-no-label">
      <input
        type="radio"
        name="registerDeliveryAddress"
        :value="false"
        v-model="register"
        data-test="register-no-input"
      />
      いいえ
    </label>
    <label data-test="register-yes-label">
      <input
        type="radio"
        name="registerDeliveryAddress"
        :value="true"
        v-model="register"
        data-test="register-yes-input"
      />
      はい
    </label>
  </fieldset>
</template>
