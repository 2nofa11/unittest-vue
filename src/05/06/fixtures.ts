export const deliveryAddresses = [
  {
    id: "address_id_xxxx",
    value: "address_id_xxxx",
    children: "〒167-0051 東京都杉並区荻窪1-00-00",
  },
  {
    id: "address_id_yyyy",
    value: "address_id_yyyy",
    children: "〒150-0041 東京都渋谷区神南2-1-1",
  },
];
