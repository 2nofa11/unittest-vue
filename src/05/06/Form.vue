<script setup lang="ts">
import { ref, computed } from "vue";
import ContactNumber from "./ContactNumber.vue";
import DeliveryAddress from "./DeliveryAddress.vue";
import PastDeliveryAddress from "./PastDeliveryAddress.vue";
import RegisterDeliveryAddress from "./RegisterDeliveryAddress.vue";

export interface AddressOption {
  id: string;
  value: string;
  children: string;
}

const props = defineProps<{
  deliveryAddresses?: AddressOption[];
}>();

const emit = defineEmits<{
  (e: "submit", values: Record<string, any>): void;
}>();

const phoneNumber = ref("");
const name = ref("");
const registerNewAddress = ref<boolean | undefined>(undefined); // Initially undefined
const selectedPastAddressValue = ref("");
const postalCode = ref("");
const prefecture = ref("");
const municipalities = ref("");
const streetAddress = ref("");

const hasPastAddresses = computed(
  () => (props.deliveryAddresses?.length || 0) > 0
);

const isPastAddressDisabled = computed(
  () => registerNewAddress.value === undefined
);

const handleFormSubmit = (event: Event) => {
  const formData = new FormData(event.target as HTMLFormElement);
  const values: { [k: string]: unknown } = {};
  formData.forEach((value, key) => (values[key] = value));
  emit("submit", values); // submit イベントを発行
};
</script>

<template>
  <form @submit.prevent="handleFormSubmit" data-test="form">
    <h2 data-test="form-heading">お届け先情報の入力</h2>

    <ContactNumber
      v-model:phoneNumber="phoneNumber"
      v-model:name="name"
      data-test="contact-number-component"
    />

    <template v-if="hasPastAddresses">
      <RegisterDeliveryAddress
        v-model="registerNewAddress"
        data-test="register-delivery-address-component"
      />
      <DeliveryAddress
        v-if="registerNewAddress === true"
        title="新しいお届け先"
        v-model:postalCode="postalCode"
        v-model:prefecture="prefecture"
        v-model:municipalities="municipalities"
        v-model:streetAddress="streetAddress"
        data-test="new-delivery-address-component"
      />
      <PastDeliveryAddress
        v-else
        :disabled="isPastAddressDisabled"
        :options="props.deliveryAddresses || []"
        v-model="selectedPastAddressValue"
        data-test="past-delivery-address-component"
      />
    </template>
    <template v-else>
      <DeliveryAddress
        v-model:postalCode="postalCode"
        v-model:prefecture="prefecture"
        v-model:municipalities="municipalities"
        v-model:streetAddress="streetAddress"
        data-test="default-delivery-address-component"
      />
    </template>

    <hr />
    <div>
      <button type="submit" data-test="submit-button">
        注文内容の確認へ進む
      </button>
    </div>
  </form>
</template>
