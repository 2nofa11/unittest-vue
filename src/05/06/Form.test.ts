import { mount, VueWrapper } from "@vue/test-utils";
import ContactNumber from "./ContactNumber.vue";
import DeliveryAddress from "./DeliveryAddress.vue";
import { deliveryAddresses } from "./fixtures";
import Form from "./Form.vue";
import PastDeliveryAddress from "./PastDeliveryAddress.vue";
import RegisterDeliveryAddress from "./RegisterDeliveryAddress.vue";

async function inputContactNumber(
  wrapper: VueWrapper<any>,
  inputValues = {
    name: "田中 太郎",
    phoneNumber: "000-0000-0000",
  }
) {
  const contactWrapper = wrapper.findComponent(ContactNumber);
  await contactWrapper
    .find('[data-test="contact-number-input"]')
    .setValue(inputValues.phoneNumber);
  await contactWrapper
    .find('[data-test="contact-name-input"]')
    .setValue(inputValues.name);
  return inputValues;
}

async function inputDeliveryAddress(
  wrapper: VueWrapper<any>,
  inputValues = {
    postalCode: "167-0051",
    prefecture: "東京都",
    municipalities: "杉並区荻窪1",
    streetAddress: "00-00",
  }
) {
  const addressWrapper = wrapper.findComponent(DeliveryAddress);
  await addressWrapper
    .find('[data-test="postal-code-input"]')
    .setValue(inputValues.postalCode);
  await addressWrapper
    .find('[data-test="prefecture-input"]')
    .setValue(inputValues.prefecture);
  await addressWrapper
    .find('[data-test="municipalities-input"]')
    .setValue(inputValues.municipalities);
  await addressWrapper
    .find('[data-test="street-address-input"]')
    .setValue(inputValues.streetAddress);
  return inputValues;
}

async function clickSubmit(wrapper: VueWrapper<any>) {
  await wrapper.find('[data-test="submit-button"]').trigger("submit");
}

function expectSubmitEventPayload(
  wrapper: VueWrapper<any>,
  expectedPayload: Record<string, any>
) {
  const emitted = wrapper.emitted("submit");
  expect(emitted).toHaveLength(1);
  expect(emitted![0]).toEqual([expect.objectContaining(expectedPayload)]);
}

describe("Form.vue", () => {
  describe("過去のお届け先がない場合", () => {
    test("お届け先入力欄がある", () => {
      const wrapper = mount(Form);
      expect(wrapper.findComponent(ContactNumber).exists()).toBe(true);
      expect(wrapper.findComponent(DeliveryAddress).exists()).toBe(true);
    });

    test("入力・送信すると、入力内容が送信される", async () => {
      const wrapper = mount(Form, {
        props: { deliveryAddresses: [] },
      });
      const contactNumber = await inputContactNumber(wrapper);
      const deliveryAddress = await inputDeliveryAddress(wrapper);
      await wrapper.find('[data-test="form"]').trigger("submit"); // Trigger submit on form

      expectSubmitEventPayload(wrapper, {
        ...contactNumber,
        ...deliveryAddress,
      });
    });

    test("Snapshot", () => {
      const wrapper = mount(Form, { props: { deliveryAddresses: [] } });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe("過去のお届け先がある場合", () => {
    test("設問に答えるまで、お届け先を選べない", () => {
      const wrapper = mount(Form, { props: { deliveryAddresses } });
      expect(wrapper.findComponent(RegisterDeliveryAddress).exists()).toBe(
        true
      );
      const pastAddressWrapper = wrapper.findComponent(PastDeliveryAddress);
      expect(pastAddressWrapper.exists()).toBe(true);

      expect((pastAddressWrapper.element as HTMLFieldSetElement).disabled).toBe(
        true
      );
    });

    test("「いいえ」を選択・入力・送信すると、入力内容が送信される", async () => {
      const wrapper = mount(Form, {
        props: { deliveryAddresses },
      });

      await wrapper.find('[data-test="register-no-input"]').setValue();

      const pastAddressWrapper = wrapper.findComponent(PastDeliveryAddress);
      const select = pastAddressWrapper.find("select");
      await select.setValue(deliveryAddresses[1].value);

      const inputValues = await inputContactNumber(wrapper);
      await wrapper.find('[data-test="form"]').trigger("submit");

      expectSubmitEventPayload(wrapper, {
        ...inputValues,
        pastDeliveryAddress: deliveryAddresses[1].value,
      });
    });

    test("「はい」を選択・入力・送信すると、入力内容が送信される", async () => {
      const wrapper = mount(Form, {
        props: { deliveryAddresses },
      });

      await wrapper.find('[data-test="register-yes-input"]').setValue();

      expect(wrapper.findComponent(DeliveryAddress).exists()).toBe(true);

      const contactNumber = await inputContactNumber(wrapper);
      const deliveryAddress = await inputDeliveryAddress(wrapper);
      await wrapper.find('[data-test="form"]').trigger("submit");

      expectSubmitEventPayload(wrapper, {
        ...contactNumber,
        ...deliveryAddress,
      });
    });

    test("Snapshot", () => {
      const wrapper = mount(Form, { props: { deliveryAddresses } });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
