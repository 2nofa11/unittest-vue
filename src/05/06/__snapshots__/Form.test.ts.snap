// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.vue 過去のお届け先がある場合 Snapshot 1`] = `
"<form data-test="form">
  <h2 data-test="form-heading">お届け先情報の入力</h2>
  <fieldset data-test="contact-number-component">
    <legend data-test="contact-legend">連絡先</legend>
    <div><label data-test="contact-number-label"> 電話番号 <input type="tel" name="phoneNumber" data-test="contact-number-input"></label></div>
    <div><label data-test="contact-name-label"> お名前 <input type="text" name="name" data-test="contact-name-input"></label></div>
  </fieldset>
  <fieldset data-test="register-delivery-address-component">
    <legend data-test="register-delivery-address-legend"> 新しいお届け先を登録しますか？ </legend><label data-test="register-no-label"><input type="radio" name="registerDeliveryAddress" data-test="register-no-input" value="false"> いいえ </label><label data-test="register-yes-label"><input type="radio" name="registerDeliveryAddress" data-test="register-yes-input" value="true"> はい </label>
  </fieldset>
  <fieldset disabled="" style="opacity: 0.3;" data-test="past-delivery-address-component">
    <legend data-test="past-delivery-address-legend">過去のお届け先</legend><select name="pastDeliveryAddress" data-test="past-delivery-address-select">
      <option data-test="past-delivery-address-option" value="address_id_xxxx">〒167-0051 東京都杉並区荻窪1-00-00</option>
    </select>
  </fieldset>
  <hr>
  <div><button type="submit" data-test="submit-button"> 注文内容の確認へ進む </button></div>
</form>"
`;

exports[`Form.vue 過去のお届け先がない場合 Snapshot 1`] = `
"<form data-test="form">
  <h2 data-test="form-heading">お届け先情報の入力</h2>
  <fieldset data-test="contact-number-component">
    <legend data-test="contact-legend">連絡先</legend>
    <div><label data-test="contact-number-label"> 電話番号 <input type="tel" name="phoneNumber" data-test="contact-number-input"></label></div>
    <div><label data-test="contact-name-label"> お名前 <input type="text" name="name" data-test="contact-name-input"></label></div>
  </fieldset>
  <fieldset data-test="default-delivery-address-component">
    <legend data-test="delivery-address-legend">お届け先</legend>
    <div><label data-test="postal-code-label"> 郵便番号 <input type="text" name="postalCode" placeholder="167-0051" data-test="postal-code-input"></label></div>
    <div><label data-test="prefecture-label"> 都道府県 <input type="text" name="prefecture" placeholder="東京都" data-test="prefecture-input"></label></div>
    <div><label data-test="municipalities-label"> 市区町村 <input type="text" name="municipalities" placeholder="杉並区荻窪1" data-test="municipalities-input"></label></div>
    <div><label data-test="street-address-label"> 番地番号 <input type="text" name="streetAddress" placeholder="00-00" data-test="street-address-input"></label></div>
  </fieldset>
  <hr>
  <div><button type="submit" data-test="submit-button"> 注文内容の確認へ進む </button></div>
</form>"
`;
